File: contents/bubble_sort/code/scala/bubble_sort.scala
object BubbleSort {

  def bubbleUpLargestToTheEnd(list: List[Int]): List[Int] =
    list match {
      case first :: second :: tail if first > second => second :: bubbleUpLargestToTheEnd(first :: tail)
      case first :: second :: tail => first :: bubbleUpLargestToTheEnd(second :: tail)
      case _ => list
    }

  def bubbleSort(list: List[Int]): List[Int] =
    bubbleUpLargestToTheEnd(list) match {
      case unsorted :+ largest => bubbleSort(unsorted) :+ largest
      case _ => list
    }


  def main(args: Array[String]): Unit = {
    val unsorted = List(9, 2, 0, 5, 3, 8, 1, 9, 4, 0, 7, 0, 9, 9, 0)

    println("Unsorted list is " + unsorted)
    println("  Sorted list is " + bubbleSort(unsorted))
  }
}

File: contents/bubble_sort/code/lua/bubble_sort.lua
function bubble_sort(arr)
  for i = 1,#arr-1 do
    for j = 1,#arr-1 do
      if arr[j] > arr[j+1] then
        arr[j], arr[j+1] = arr[j+1], arr[j]
      end
    end
  end
end

local arr = {1, 45, 756, 4569, 56, 3, 8, 5, -10, -4}
print(("Unsorted array: {%s}"):format(table.concat(arr,", ")))

bubble_sort(arr)

print(("Sorted array: {%s}"):format(table.concat(arr,", ")))
